Add RSS feed that works without OAuth.

punchagan opened the pull request:
We use a random token generated per-user to authenticate the user for
obtaining a feed. This feed can now be used by feed aggregators without
having to be authenticated using OAuth.

Thanks @davidalbert!

punchagan commented:
![image](https://cloud.githubusercontent.com/assets/315678/7247795/f84af302-e82a-11e4-834b-c97c687b5ac0.png)

davidbalbert commented:
@punchagan This looks reasonable (again, I have very little Python experience, so you probably shouldn't take my approval to mean this is good to go). A few thoughts:

1. It doesn't look like you're generating tokens for the existing users. This is something you probably want to do.
2. What is the default token in the migration for?
3. You require that tokens are unique, but you're not checking that your generated token is unique. The odds of generating a duplicate token are incredibly low, so maybe this is ok, but I think the cost of ensuring that you generate a unique token is also pretty low (basically just a while loop).
4. You probably want to add a database index on the `token` field.

punchagan commented:
Thanks for the review @davidbalbert. I had never really bothered into how Django migrations are done, and mostly trusted/depended on `South` to do the right things. I learnt something about mirgations today, thanks to your questions!

I use a `default` value on the `Model`, so that all new users automatically get a new token. A default value is in the migration script, since by default the field is expected to be non NULL.   

The whole has now been split into 3 steps.

  1. `schemamigration` with no unique constraint on the token field,
  2. `datamigration` ensuring that all existing users have unique tokens
  3. `schemamigration` that adds uniqueness constraint to token field.

Also, I changed the field from a `CharField` to a `SlugField` so that it is indexed.

I was being lazy to check for the uniqueness of the token; I have now added it.

FiloSottile commented:
+1

Would love to have this in production!

sursh commented:
Yeah, let's get this out the door! The code looks good to me. I can test it on staging. 

From a user experience perspective, I think there's a little to clean up. Let's make it clear to the reader how to find the RSS feed, and what will be displayed in it. 

Some specific suggestions: 

- Remove the display of the token. I thought when looking at the screenshots that they had to copy and paste it to do something with it, but in looking at the code it looks like they just have to click that link which is already constructed with their token. 
- Put the link on the main page (with all the posts) rather than the profile page. It seems clearer because that's the page you go do to read and browse the posts, whereas the settings page is where you go to do edit your own settings. 
- Perhaps add a line in the About page.

sursh commented on home/migrations/0013_random_tokens.py:
```diff
@@ -0,0 +1,124 @@
+                    return token
+
+        for hacker in Hacker.objects.all():
+            hacker.token = get_random_unique_token(Hacker)
```
Bug?

sursh commented:
Testing this on staging: 
- running a schemamigration (`heroku run python manage.py schemamigration home --auto`) returns `Nothing seems to have changed.` Not what I expected, but OK.... 
- running the migration (`heroku run python manage.py migrate home`) throws an error. I think it's from line 30 above. But I'm a little puzzled as I'm pretty sure this migration script is autogenerated. 

```Running migrations for home:
 - Migrating forwards to 0014_auto__add_unique_hacker_token.
 > home:0012_auto__add_field_hacker_token
 > home:0013_random_tokens
Error in migration: home:0013_random_tokens
TypeError: get_random_unique_token() takes no arguments (1 given)```

sursh commented:
- I have reminded myself of what `schemamigration` actually does and see why it didn't do anything when I ran it :)
- I took the liberty of fixing the bug in the migration script and it runs nicely on staging. Feel free to play with it. 
- I think we will be ready to :shipit: after ironing out the user experience things. @punchagan are you ok with 
1) adding an RSS link/icon to the bottom of the main page?
2) small changes to the copy & layout of the settings page to make it clearer what the feed & token are for?

punchagan commented:
@sursh Sure! :+1:  But, I'm not sure I will be able to do it before the weekend. If you want to push it out sooner than that, feel free to make the changes!

punchagan commented:
I finally took a stab at checking off the remaining two items. @sursh, feel free to make any changes you feel are necessary!

punchagan commented:
![blag-rss](https://cloud.githubusercontent.com/assets/315678/10685006/9a5101aa-7971-11e5-803e-d057ba81488c.jpg)

punchagan commented:
![image](https://cloud.githubusercontent.com/assets/315678/10867380/d36bac78-8084-11e5-9bac-7eaa23de1ae1.png)

I pushed a few more changes. 
- Completely remove any display of the auth token to the user. 
- Minor changes to the profile page -- title, fix empty blog list, add blog button text.
- Clean up view for feed. 
- Fix wrong link attribute (example.com) for the feed. 
  I removed the `sites` app to fix this. Our app isn't really supporting multiple different sites, and that is clear from the fact that we had just one SITE_ID. The sites features weren't being used (correctly) and can safely be removed. 

These changes are deployed on staging. @sursh if you can take a quick look, I can deploy it to production in case you don't have time to do so.

punchagan commented:
Superseded by #80
